Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -1260,28 +1262,54 @@ size_t MatrixWorkspace::binIndexOf(const double xValue,
     throw std::out_of_range("MatrixWorkspace::binIndexOf - X value greater"
                             " than highest in current range.");
   }
-  size_t hops;
+
+  // Check if the workspace is a histogram workspace (len(x) = len(y) + 1 for a
+  // histogram workspace)
+  auto const ySize = this->y(index).size();
+  auto const xSize = xValues.size();
+  bool const isHistogram = xSize == ySize + 1;
+
+  std::size_t hops;
   if (ascendingOrder) {
-    auto lowit = std::lower_bound(xValues.cbegin(), xValues.cend(), xValue);
+    auto lowerIter =
+        std::lower_bound(xValues.cbegin(), xValues.cend(), xValue - tolerance);
+    auto const upperIter =
+        std::lower_bound(xValues.cbegin(), xValues.cend(), xValue + tolerance);
+
     // If we are pointing at the first value then that means we still want to be
     // in the first bin
-    if (lowit == xValues.cbegin()) {
-      ++lowit;
-    }
-    hops = std::distance(xValues.cbegin(), lowit);
+    if (isHistogram && lowerIter == xValues.cbegin())
+      ++lowerIter;
+    else if (!isHistogram && lowerIter == xValues.cend())
+      --lowerIter;
+
+    hops = std::distance(xValues.cbegin(), lowerIter);
+
+    // If the tolerance makes no difference in the chosen iterator
```

Comment:
Could you help me understand what this comment means and what the next if is doing? I don't quite follow it...

It is also generally better do avoid comparing double values exactly so it might be better to say `tolerance > 0.0` here rather than `!= 0`.

Reply:
I'm finding the code a little difficult to follow as we have many `isHistogram` checks and 2 places that seem to have the same check for `tolerance != 0.0`. Would it be possible just to have a single check for `isHistogram` and the `if` part deals with that and the `else` deals with point data?

Reply:
I think I will change this to make it easier to read. If the workspace is for point data and the xValue passed is not stored in the xValues (+- tolerance), then this should probably throw (rather than having the if statement above)